### ISSUE TITLE
[Bug] failed to run custom build command for `dart-sys v3.1.12`

### ISSUE THREAD

**Issue by WangJiaJun8922** [fzyzcjy/flutter_rust_bridge#1021]
### Describe the bug

Hi, it's me again. 😂😂😂

flutter_rust_bridge = "1.62.0"
cargo-ndk = "2.12.6"

It's strange! I followed the tutorial to integrate Android again today, re-entered the commands in the tutorial, so the libraries should have been updated! But Android is unable to compile now.

**`fatal error: 'assert.h' file not found`**

```bash
 Compiling bzip2 v0.4.4
   Compiling allo-isolate v0.1.14
   Compiling zstd v0.11.2+zstd.1.5.2
   Compiling zip v0.6.4
   Compiling dart-sys v3.1.12
The following warnings were emitted during compilation:

warning: INFO: build log is located at: `/Users/ken/codes/hs_project_root/test_res/test_2/rust/target/armv7-linux-androideabi/debug/build/dart-sys-46e53f82cb928d91/out/build.log`

error: failed to run custom build command for `dart-sys v3.1.12`

Caused by:
  process didn't exit successfully: `/Users/ken/codes/hs_project_root/test_res/test_2/rust/target/debug/build/dart-sys-9bd5cd6c06da5acf/build-script-build` (exit status: 101)
  --- stdout
  cargo:warning=INFO: build log is located at: `/Users/ken/codes/hs_project_root/test_res/test_2/rust/target/armv7-linux-androideabi/debug/build/dart-sys-46e53f82cb928d91/out/build.log`

  --- stderr
  /Users/ken/.cargo/registry/src/github.com-1ecc6299db9ec823/dart-sys-3.1.12/dart-sdk/include/dart_api.h:24:10: fatal error: 'assert.h' file not found
  thread 'main' panicked at 'ERROR: Failed to generate dart_api_dl binding: ClangDiagnostic("/Users/ken/.cargo/registry/src/github.com-1ecc6299db9ec823/dart-sys-3.1.12/dart-sdk/include/dart_api.h:24:10: fatal error: 'assert.h' file not found\n")', /Users/ken/.cargo/registry/src/github.com-1ecc6299db9ec823/dart-sys-3.1.12/build.rs:666:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-02-02T08:10:26Z INFO  cargo_ndk::cli] If the build failed due to a missing target, you can run this command:
[2023-02-02T08:10:26Z INFO  cargo_ndk::cli] 
[2023-02-02T08:10:26Z INFO  cargo_ndk::cli]     rustup target install armv7-linux-androideabi

rust on  master [?] is 📦 v0.1.0 via 🦀 v1.65.0 took 14s 
```

However, when I executed `flutter clean`   `cargo-clean` on my previous project and recompiled, there were no problems. There was no output of "Compiling Dart-sys" during the compilation.

So I checked the Cargo.lock, it should be 1.62.0, which added the library `dart-sys` compared to 1.61.1.

- 1.62.0
```toml
[[package]]
name = "flutter_rust_bridge"
version = "1.62.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a943fe1e3cf0477300f919946088f12b16e79daad9a0cfb4f87f0d63e908bf80"
dependencies = [
 "allo-isolate",
 "anyhow",
 "build-target",
 "bytemuck",
 "cc",
 "chrono",
 "console_error_panic_hook",
 "dart-sys",
 "flutter_rust_bridge_macros",
 "js-sys",
 "lazy_static",
 "libc",
 "log",
 "parking_lot",
 "threadpool",
 "wasm-bindgen",
 "web-sys",
]
```

- 1.61.1
```toml
[[package]]
name = "flutter_rust_bridge"
version = "1.61.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8079119bbe8fb63d7ebb731fa2aa68c6c8375f4ac95ca26d5868e64c0f4b9244"
dependencies = [
 "allo-isolate",
 "anyhow",
 "build-target",
 "bytemuck",
 "cc",
 "chrono",
 "console_error_panic_hook",
 "flutter_rust_bridge_macros",
 "js-sys",
 "lazy_static",
 "libc",
 "log",
 "parking_lot",
 "threadpool",
 "wasm-bindgen",
 "web-sys",
]
```

There must have been changes last week.
- frb_rust/Cargo.toml
``` rust
@@ -24,7 +24,7 @@ allo-isolate = { workspace = true }
libc = "0.2"
parking_lot = { version = "0.12.1" }
threadpool = { version = "1.8.1" }
- dart-sys = { git="https://github.com/rogurotus/dart-sys.git", branch="whatif", features = ["dart_api_dl"] }
+ dart-sys = { version = "3.1.12", features = ["dart_api_dl"] }

[target.'cfg(target_family = "wasm")'.dependencies]
console_error_panic_hook = { version = "0.1.7", optional = true }
```

The dart-sys was updated to 4.0 just a few hours ago. and I don't know what caused the issue, but I can only roll back the version by running `cargo update --package flutter_rust_bridge --precise 1.61.1`

### Codegen logs with `RUST_LOG=debug` environment variable

```shell
Compiling bzip2 v0.4.4
   Compiling allo-isolate v0.1.14
   Compiling zstd v0.11.2+zstd.1.5.2
   Compiling zip v0.6.4
   Compiling dart-sys v3.1.12
The following warnings were emitted during compilation:

warning: INFO: build log is located at: `/Users/ken/codes/hs_project_root/test_res/test_2/rust/target/armv7-linux-androideabi/debug/build/dart-sys-46e53f82cb928d91/out/build.log`

error: failed to run custom build command for `dart-sys v3.1.12`

Caused by:
  process didn't exit successfully: `/Users/ken/codes/hs_project_root/test_res/test_2/rust/target/debug/build/dart-sys-9bd5cd6c06da5acf/build-script-build` (exit status: 101)
  --- stdout
  cargo:warning=INFO: build log is located at: `/Users/ken/codes/hs_project_root/test_res/test_2/rust/target/armv7-linux-androideabi/debug/build/dart-sys-46e53f82cb928d91/out/build.log`

  --- stderr
  /Users/ken/.cargo/registry/src/github.com-1ecc6299db9ec823/dart-sys-3.1.12/dart-sdk/include/dart_api.h:24:10: fatal error: 'assert.h' file not found
  thread 'main' panicked at 'ERROR: Failed to generate dart_api_dl binding: ClangDiagnostic("/Users/ken/.cargo/registry/src/github.com-1ecc6299db9ec823/dart-sys-3.1.12/dart-sdk/include/dart_api.h:24:10: fatal error: 'assert.h' file not found\n")', /Users/ken/.cargo/registry/src/github.com-1ecc6299db9ec823/dart-sys-3.1.12/build.rs:666:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-02-02T08:10:26Z INFO  cargo_ndk::cli] If the build failed due to a missing target, you can run this command:
[2023-02-02T08:10:26Z INFO  cargo_ndk::cli] 
[2023-02-02T08:10:26Z INFO  cargo_ndk::cli]     rustup target install armv7-linux-androideabi

rust on  master [?] is 📦 v0.1.0 via 🦀 v1.65.0 took 14s
```

**Comment by kingfisherphuoc** [comment]
to me, it's the same bug but different logs:
```
warning: INFO: build log is located at: `/Users/kingfisher-phuoc/DevTool/AndroidDevTool/Workspace_Flutter/goong_speed_limit/GoongSpeedLimit/native/target/aarch64-linux-android/debug/build/dart-sys-65b67597f37148e9/out/build.log`

error: failed to run custom build command for `dart-sys v3.1.12`

Caused by:
  process didn't exit successfully: `/Users/kingfisher-phuoc/DevTool/AndroidDevTool/Workspace_Flutter/goong_speed_limit/GoongSpeedLimit/native/target/debug/build/dart-sys-5df1e9e54eee9cdb/build-script-build` (exit status: 101)
  --- stdout
  cargo:warning=INFO: build log is located at: `/Users/kingfisher-phuoc/DevTool/AndroidDevTool/Workspace_Flutter/goong_speed_limit/GoongSpeedLimit/native/target/aarch64-linux-android/debug/build/dart-sys-65b67597f37148e9/out/build.log`

  --- stderr
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/cdefs.h:882:2: error: Unsupported architecture
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/machine/_types.h:36:2: error: architecture not supported
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:55:9: error: unknown type name '__int64_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:56:9: error: unknown type name '__int32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:57:9: error: unknown type name '__int32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:60:9: error: unknown type name '__uint32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:61:9: error: unknown type name '__uint32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:62:9: error: unknown type name '__uint64_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:68:9: error: unknown type name '__darwin_natural_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:70:9: error: unknown type name '__uint16_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:71:9: error: unknown type name '__int64_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:72:9: error: unknown type name '__int32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:73:9: error: unknown type name '__uint32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:74:9: error: unknown type name '__int32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:75:9: error: unknown type name '__uint32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:76:9: error: unknown type name '__uint32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/_types.h:43:9: error: unknown type name '__uint32_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types/_wchar_t.h:34:9: error: unknown type name '__darwin_wchar_t'
  /Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/machine/types.h:39:2: error: architecture not supported
  fatal error: too many errors emitted, stopping now [-ferror-limit=]
  thread 'main' panicked at 'ERROR: Failed to generate dart_api_dl binding: ClangDiagnostic("/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/cdefs.h:882:2: error: Unsupported architecture\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/machine/_types.h:36:2: error: architecture not supported\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:55:9: error: unknown type name '__int64_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:56:9: error: unknown type name '__int32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:57:9: error: unknown type name '__int32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:60:9: error: unknown type name '__uint32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:61:9: error: unknown type name '__uint32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:62:9: error: unknown type name '__uint64_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:68:9: error: unknown type name '__darwin_natural_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:70:9: error: unknown type name '__uint16_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:71:9: error: unknown type name '__int64_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:72:9: error: unknown type name '__int32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:73:9: error: unknown type name '__uint32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:74:9: error: unknown type name '__int32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:75:9: error: unknown type name '__uint32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types.h:76:9: error: unknown type name '__uint32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/_types.h:43:9: error: unknown type name '__uint32_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/sys/_types/_wchar_t.h:34:9: error: unknown type name '__darwin_wchar_t'\n/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/usr/include/machine/types.h:39:2: error: architecture not supported\nfatal error: too many errors emitted, stopping now [-ferror-limit=]\n")', /Users/kingfisher-phuoc/.cargo/registry/src/github.com-1ecc6299db9ec823/dart-sys-3.1.12/build.rs:666:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
[2023-02-02T15:31:28Z INFO  cargo_ndk::cli] If the build failed due to a missing target, you can run this command:
[2023-02-02T15:31:28Z INFO  cargo_ndk::cli] 
[2023-02-02T15:31:28Z INFO  cargo_ndk::cli]     rustup target install aarch64-linux-android
```
any though?

**Comment by fzyzcjy** [comment]
So, does #1006's solution solve the bug?

**Comment by kingfisherphuoc** [comment]
no, it does not. I must set `flutter_rust_bridge = "=1.61.1"` in rust and other stuff in dart.

**Comment by fzyzcjy** [comment]
Hmm cc @rogurotus who implemented the good dart-sys related things, and maybe @gutenfries who updated dart-sys recently

**Comment by WangJiaJun8922** [comment]
no,  i see #1006  use `cargo-ndk = 2.12.5` to solve it.    `2.12.5, 2.12.6, 2.12.4`  I tried them all, it didn't work.
like @kingfisherphuoc say.   
roll back any `flutter_rust_bridge_xxx` to `1.61.1`. include  `Cargo.lock`

**Comment by SaamBaby** [comment]
Thanks @kingfisherphuoc , I was having the same problem, but when i rolled back flutter_rust_bridge to 1.61.0
 Everything went back to normal

**Comment by SupremeDeity** [comment]
> no, i see #1006 use `cargo-ndk = 2.12.5` to solve it. `2.12.5, 2.12.6, 2.12.4` I tried them all, it didn't work. like @kingfisherphuoc say. roll back any `flutter_rust_bridge_xxx` to `1.61.1`. include `Cargo.lock`

Just for context, while solving this issue, i **didnt** put `cargo-ndk = 2.12.5` in my `cargo.toml`. I simply used the command `cargo install cargo-ndk --version 2.12.5`

Reference:
https://cjycode.com/flutter_rust_bridge/template/setup_android.html

**Comment by WangJiaJun8922** [comment]
@SupremeDeity yes, I saw your issues yesterday, and according to your issues I tried `cargo install cargo-ndk@2.12.5` `@2.12.4` `@2.12.6`, none of which worked for me. but my other `1.61.1` projects can build success, so I guess it may be the fault of `dart-sys`.

**Comment by SupremeDeity** [comment]
> @SupremeDeity yes, I saw your issues yesterday, and according to your issues I tried `cargo install cargo-ndk@2.12.5` `@2.12.4` `@2.12.6`, none of which worked for me. but my other `1.61.1` projects can build success, so I guess it may be the fault of `dart-sys`.

At this point i dont even know what is happening because i tried using frb 1.61.1 and it wasnt working for me. I had to waste like 2-3 days. Just using ndk 2.12.5 worked

**Comment by gutenfries** [comment]
@fzyzcjy:

@rogurotus and I _ARE AWARE_ that certain versions of https://github.com/dart-sys/dart-sys are broken and we are working through different difficulties.

I would ask that you yank any versions of FRB that implement Dart-sys <`v4.0.2` and please test and review extensively before actually merging changes concerning the inclusion of Dart-sys.

I _believe_ that `v4.0.2` is stable and platform-independent, however I think more extensive testing should be done before merging the use of Dart-sys into FRB again.

please also read: #1023

**Comment by fzyzcjy** [comment]
@gutenfries 

Given that the current frb version is broken, I would like to merge the new updates of dart-sys (given that it cannot be worse than broken) to at least master. But yes, according to your advice, I will not release new version after merge, but put it to master and see whether it resolves the bugs here.

**Comment by fzyzcjy** [comment]
#1023 is already merged. Could you please check whether the bugs related to dart-sys is resolved now?

**Comment by gutenfries** [comment]
> #1023 is already merged. Could you please check whether the bugs related to dart-sys are resolved now?

@fzyzcjy, it has been a few days since #1023 and I haven't found any issues, and there do not appear to be any newly opened issues related to Dart-sys. I think that we (finally lol) have a good build of dart-sys.

In the near-ish future, @rogurotus and/or I (@gutenfries) will open another PR/issue when I have refactored Dart-sys to support 100% opt-in style bindings, as is used in web-sys (https://github.com/rustwasm/wasm-bindgen/tree/main/crates/web-sys).

This should future PR should present no breaking changes, and will result in faster build times and smaller bundles for FRB. :)

**Comment by gutenfries** [comment]
(@fzyzcjy) related/duplicate issue: #1022

**Comment by GregoryConrad** [comment]
@WangJiaJun8922  This issue is fixed for me in 1.62.1!

**Comment by fzyzcjy** [comment]
Great job! 🎉

**Comment by AlexV525** [comment]
I'm facing this issue when trying to build with FRB 1.64.1. See the "Execute build scripts" step in this job: https://github.com/AstroxNetwork/agent_dart/actions/runs/4229506512/jobs/7345944137

```
......
   Compiling dart-sys v4.0.2
error: failed to run custom build command for `dart-sys v4.0.2`
Error: failed to run custom build command for `dart-sys v4.0.2`
```

**Comment by fzyzcjy** [comment]
@AlexV525 Hmm the CI in this repo works well, so maybe it can be solved by mimicking it? https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/.github/workflows/ci.yaml

**Comment by AlexV525** [comment]
> https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/.github/workflows/ci.yaml

Thanks. The example helps, changing the `cargo build` to `cargo ndk ... build` solves this.
